### ISSUE TITLE
Compile time is painfully slow

### ISSUE THREAD

**Issue by jpike88** [angular/angular-cli#24922]
### Command

build, serve

### Description

The ng build/serve compilation times of angular dwarf any other bundler/framework I know of and this is a real drag on my team's development time.

I am wondering if there is perhaps a way to just disable type checking or other things that may be large contributing factors to build time, or some sort of alternative being explored?

Can you also explain a little bit about what % the bottlenecks are right now? It would be good to understand things a little better.

### Describe the solution you'd like

A way faster compilation time

### Describe alternatives you've considered

There aren't any, because since the linker change in Angular 13, I can't use a 3rd party bundler anymore.

**Comment by JoostK** [comment]
> There aren't any, because since the linker change in Angular 13, I can't use a 3rd party bundler anymore.

You certainly could! The linker is just a Babel plugin that can be incorporated into any build pipeline.

**Comment by alan-agius4** [comment]
> The ng build/serve compilation times of angular dwarf any other bundler/framework I know of and this is a real drag on my team's development time.

What are your build times? Did you try using the experimental esbuild based builder? 

> I am wondering if there is perhaps a way to just disable type checking or other things that may be large contributing factors to build time, or some sort of alternative being explored?

Have you profiled the build to verify the type-checking is the culprit here?

**Comment by jpike88** [comment]
Build times are about 30 seconds on first build. Esbuild doesn't seem to make much of a difference.

I've profiled my build running `node --cpu-prof node_modules/.bin/ng build`, here it is:

[profile.zip](https://github.com/angular/angular-cli/files/11108407/profile.zip)

It's hard to interpret exactly why it's slow, all I know is that it's slow. And It's not a simple stack to interpret, I have no idea how to answer your question even with the profiles.

<img width="694" alt="Screenshot 2023-03-30 at 1 21 11 pm" src="https://user-images.githubusercontent.com/9585787/228747251-c0c41355-9c0c-420f-984c-c56dadcd68fd.png">

**Comment by JoostK** [comment]
Those are profiles from a production build, right?

**Comment by jpike88** [comment]
nope, development. our default build is our development build

**Comment by jpike88** [comment]
or do you mean a production build of angular-cli? If so then yes, v15.1.5

**Comment by jpike88** [comment]
whoops! our ng build and ng serve are configured a little differently for their defaults, I forgot. Here is it properly optimizaed for development, which is closer to the actual speed of build on an M1 Mac (~18 seconds):

Config:
```
{
							"assets": [
								{
									"glob": "**/*",
									"input": "src/assets/",
									"output": "/"
								},
								{
									"glob": "**/*",
									"input": "node_modules/ngx-extended-pdf-viewer/assets/",
									"output": "/assets/"
								}
							],
							"optimization": false,
							"outputHashing": "none",
							"sourceMap": {
								"scripts": true,
								"styles": true,
								"vendor": false,
								"hidden": false
							},
							"namedChunks": false,
							"aot": true,
							"extractLicenses": false,
							"statsJson": false,
							"progress": true,
							"vendorChunk": true,
							"buildOptimizer": false
						}
```

attached result
[CPU.20230330.183927.62264.0.001.cpuprofile.zip](https://github.com/angular/angular-cli/files/11110973/CPU.20230330.183927.62264.0.001.cpuprofile.zip)

So not 30 seconds, about 18.

This to me is still uncomfortably slow, both on boot up and waiting from 4 up to another 15 seconds each update is punishing. where is the bottleneck to you in the latest cpuprofile I have?

**Comment by JoostK** [comment]
Do you have profiles of _incremental_ compilations, not the initial one after startup?

The first _and second_ build are known to be more expensive; only after the second rebuild can incremental type-checking be used due to some technicalities with how template type-checking works. So I'm interested in profiles of rebuilds _after_ the second one, ideally with a no-op change (e.g. introducing a `console.log`)

**Comment by jpike88** [comment]
Is there any intent or initiative to improve the speed of the first two builds? I do them often enough that it’s far from ideal. Esbuild had a small but insubstantial effect, is there room to optimise further?

**Comment by JoostK** [comment]
Not that I'm aware of at the moment. There's ideas like angular/angular#43131 and angular/angular#43165 that are relevant here.

**Comment by jpike88** [comment]
It would be good for this to be higher on the list of priorities. So much value is lost in waiting during those precious seconds, it breaks flow.

**Comment by jpike88** [comment]
[CPU.20230331.113255.80503.0.001.cpuprofile.zip](https://github.com/angular/angular-cli/files/11118176/CPU.20230331.113255.80503.0.001.cpuprofile.zip)

A first time run with angular 16 takes over a minute. On an M1 Mac chip.

Skimming it, the phases look like (please correct me):

| Phase | Timing |
|--------|--------|
| Angular Tools/Webpack/Iny initialisation/setup | 1.2 seconds |
| Typescript compilation | 6 seconds |
| ??? | 40 seconds |
| terminal output because I had verbose = true | 5 seconds |
| Webpack building stuff | 6 seconds |
| Webpack updating stats or something | 1 seconds |
| Writing of JS files? | 3 seconds |
| Writing of assets? | 0.5 seconds |

**Comment by JoostK** [comment]
Support for `ng serve` is being implemented in #24923; esbuild support is not yet feature complete.

The esbuild profile is as expected to me. The main difference in why the CLI with esbuild isn't nearly as fast as bare esbuild is that the TypeScript compiler is still used to parse + type-check + emit modules, given that the AOT compiler is built on top of TypeScript. esbuild is _only_ used for bundling/optimization purposes.

**Comment by jpike88** [comment]
Oh that’s great timing, didn’t see it. I’m glad that it looks like it’s getting attention anyway.

its the only thing about Angular Im not a fan of, I’ve been using it since the v1 days and love where it’s heading… thanks for your time
